### PR TITLE
Detect HTTPS ClientHello instead of manually listing the HTTPS ports +bugfix

### PR DESCRIPTION
**Pull request by honfika**
- Using upstream proxy when host is excluded, too (earlier upstream proxy was ignored in this case which is wrong)

Doneness:
- [X] Build is okay - I made sure that this change is building successfully.
- [X] No Bugs - I made sure that this change is working properly as expected. It doesn't have any bugs that you are aware of. 
- [X] Branching - If this is not a hotfix, I am making this request against develop branch 

I tried to test it carefully with HTTP site, excluded HTTPS site and non excluded HTTPS site. With and without upstream proxy. It seems to work now in all cases. I hope there is no bug... or at least better then earlier:)

### CODE REVIEW

File: Titanium.Web.Proxy/Http/HttpsTools.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+{
+    class HttpsTools
+    {
+        public static async Task<bool> IsClientHello(CustomBufferedStream clientStream)
+        {
+            //detects the HTTPS ClientHello message as it is described in the following url:
+            //https://stackoverflow.com/questions/3897883/how-to-detect-an-incoming-ssl-https-handshake-ssl-wire-format
+
+            int recordType = await clientStream.PeekByteAsync(0);
+            if (recordType == 0x80)
```
Comment (honfika) [inline Titanium.Web.Proxy/Http/HttpsTools.cs]: I did not found any website with SSLv2, so this case is not tested.

File: Titanium.Web.Proxy/RequestHandler.cs [hunk 2]
```diff
@@ -104,61 +103,87 @@ partial class ProxyServer
                         connectRequestHeaders.Add(newHeader);
                     }
 
-                    if (await CheckAuthorization(clientStreamWriter, connectRequestHeaders) == false)
+                    if (httpRemoteUri.Port == 80)
```
Comment (justcoding121) [inline Titanium.Web.Proxy/RequestHandler.cs]: HTTPS was not allowed to port 80 before for some reason before. If we don't have any problem now we can remove the check.

File: Titanium.Web.Proxy/Http/HeaderCollection.cs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Titanium.Web.Proxy.Models;
+
+namespace Titanium.Web.Proxy.Http
+{
+    public class HeaderCollection : IEnumerable<HttpHeader>
```
Comment (justcoding121) [inline Titanium.Web.Proxy/Http/HeaderCollection.cs]: :thumbsup:

File: Titanium.Web.Proxy/ProxyServer.cs [hunk 4]
```diff
@@ -172,6 +172,10 @@ public CertificateEngine CertificateEngine
         /// </summary>
         public event Func<object, SessionEventArgs, Task> BeforeResponse;
 
+        public event Func<object, TunnelConnectEventArgs, Task> TunnelConnectRequest;
+
+        public event Func<object, TunnelConnectEventArgs, Task> TunnelConnectResponse;
```
Comment (justcoding121) [inline Titanium.Web.Proxy/ProxyServer.cs]: I don't see this getting called anywhere.

Comment (honfika) [inline Titanium.Web.Proxy/ProxyServer.cs]: Yes, you are right, the response event is not called, yet. I'm still working on that.
I also plan to add some SSL information to the ConnectRequest/Response classes, line in another proxy, for example:
"A SSLv3-compatible ClientHello handshake was found. Extracted the parameters below.

Version: 3.3 (TLS/1.2)
Random: D3 FA 7E 31 4B EF 65 2E CA BA 3F AB 96 32 FE 42 53 DE 97 F9 11 B1 EC DB B9 60 B5 9F 52 63 C4 E4
"Time": 1996. 04. 25. 6:08:51
SessionID: 2C F7 B0 21 D9 83 81 A4 67 AA 4F 03 2C 4F 12 5A 33 52 F9 D6 3E 83 9F 1F 78 47 C0 8D 90 D4 AA 1E"

You can wait with the merge for that implementation, or temporary remove the TunnelConnectResponse event.

**Comment by honfika** [comment]
In the ProxyAuthorizationHelper.cs/CheckAuthorization method at this line:
`return await AuthenticateUserFunc(username, password);`
a response should be sent when the result is false.
I'm not sure what, maybe 401 Authorization Required

**Pull request merged**